From bb21a98f1ae482d6c15bb81f828546abf989e16c Mon Sep 17 00:00:00 2001
From: Arindam Biswas <quic_arinbisw@quicinc.com>
Date: Fri, 14 Feb 2025 17:30:19 +0530
Subject: [PATCH] v4l2: Add support for V4L2_PIX_FMT_QC08C format

Upstream-Status: Inappropriate [Qualcomm specific]

Signed-off-by: Arindam Biswas <quic_arinbisw@quicinc.com>
---
 sys/v4l2/ext/videodev2.h | 1 +
 sys/v4l2/gstv4l2object.c | 1 +
 2 files changed, 2 insertions(+)

diff --git a/sys/v4l2/ext/videodev2.h b/sys/v4l2/ext/videodev2.h
index 2ceb1ca..e3b4ab3 100644
--- a/sys/v4l2/ext/videodev2.h
+++ b/sys/v4l2/ext/videodev2.h
@@ -775,6 +775,7 @@ struct v4l2_pix_format {
 #define V4L2_PIX_FMT_QC08C    v4l2_fourcc('Q', '0', '8', 'C') /* Qualcomm 8-bit compressed */
 #define V4L2_PIX_FMT_QC10C    v4l2_fourcc('Q', '1', '0', 'C') /* Qualcomm 10-bit compressed */
 #define V4L2_PIX_FMT_AJPG     v4l2_fourcc('A', 'J', 'P', 'G') /* Aspeed JPEG */
+#define V4L2_PIX_FMT_QC08C    v4l2_fourcc('Q', '0', '8', 'C') /* Qualcomm 8-bit compressed */
 
 /* 10bit raw packed, 32 bytes for every 25 pixels, last LSB 6 bits unused */
 #define V4L2_PIX_FMT_IPU3_SBGGR10	v4l2_fourcc('i', 'p', '3', 'b') /* IPU3 packed 10-bit BGGR bayer */
diff --git a/sys/v4l2/gstv4l2object.c b/sys/v4l2/gstv4l2object.c
index ef9c0fc..6d922cc 100644
--- a/sys/v4l2/gstv4l2object.c
+++ b/sys/v4l2/gstv4l2object.c
@@ -156,6 +156,7 @@ static const GstV4L2FormatDesc gst_v4l2_formats[] = {
   {MAP_FMT (M420, UNKNOWN),   GST_V4L2_RAW},
 
   /* two planes -- one Y,         one Cr + Cb interleaved  */
+  {MAP_FMT (QC08C, NV12_Q08C),                  GST_V4L2_RAW},
   {MAP_FMT (NV12M, NV12),                       GST_V4L2_RAW},
   {MAP_FMT (NV12, NV12),                        GST_V4L2_RAW},
   {MAP_FMT (NV12MT, NV12_64Z32),                GST_V4L2_RAW},
-- 
2.34.1

