From ae6ea25accb8b7531aae90b6aa303a5463123f9e Mon Sep 17 00:00:00 2001
From: Naveen Mittal <namittal04@codeaurora.org>
Date: Thu, 18 Apr 2019 17:20:31 +0530
Subject: [PATCH] Expose support for DSD for 1.14.4

---
 ext/libav/gstavcodecmap.c | 28 ++++++++++++++++++++++++++++
 ext/libav/gstavdemux.c    |  2 ++
 2 files changed, 30 insertions(+)

diff --git a/ext/libav/gstavcodecmap.c b/ext/libav/gstavcodecmap.c
index 9a72033..33f731a 100644
--- a/ext/libav/gstavcodecmap.c
+++ b/ext/libav/gstavcodecmap.c
@@ -2216,6 +2216,34 @@ gst_ffmpeg_codecid_to_caps (enum AVCodecID codec_id,
           gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/G729",
           NULL);
       break;
+    case AV_CODEC_ID_DSD_LSBF:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-dsd",
+          NULL);
+      gst_caps_set_simple (caps, "lsbf", G_TYPE_BOOLEAN,
+          TRUE, "planar", G_TYPE_BOOLEAN, FALSE, NULL);
+      break;
+    case AV_CODEC_ID_DSD_MSBF:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-dsd",
+          NULL);
+      gst_caps_set_simple (caps, "lsbf", G_TYPE_BOOLEAN,
+          FALSE, "planar", G_TYPE_BOOLEAN, FALSE, NULL);
+      break;
+    case AV_CODEC_ID_DSD_LSBF_PLANAR:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-dsd",
+          NULL);
+      gst_caps_set_simple (caps, "lsbf", G_TYPE_BOOLEAN,
+          TRUE, "planar", G_TYPE_BOOLEAN, TRUE, NULL);
+      break;
+    case AV_CODEC_ID_DSD_MSBF_PLANAR:
+      caps =
+          gst_ff_aud_caps_new (context, NULL, codec_id, encode, "audio/x-dsd",
+          NULL);
+      gst_caps_set_simple (caps, "lsbf", G_TYPE_BOOLEAN,
+          FALSE, "planar", G_TYPE_BOOLEAN, TRUE, NULL);
+      break;  
     default:
       GST_DEBUG ("Unknown codec ID %d, please add mapping here", codec_id);
       break;
diff --git a/ext/libav/gstavdemux.c b/ext/libav/gstavdemux.c
index 6f3a7af..0fc2413 100644
--- a/ext/libav/gstavdemux.c
+++ b/ext/libav/gstavdemux.c
@@ -2089,6 +2089,8 @@ gst_ffmpegdemux_register (GstPlugin * plugin)
         !strcmp (in_plugin->name, "mpc") ||
         !strcmp (in_plugin->name, "mpc8") ||
         !strcmp (in_plugin->name, "ivf") ||
+        !strcmp (in_plugin->name, "dsf") ||
+        !strcmp (in_plugin->name, "iff") ||
         !strcmp (in_plugin->name, "brstm") ||
         !strcmp (in_plugin->name, "bfstm") || !strcmp (in_plugin->name, "gif"))
       rank = GST_RANK_MARGINAL;
-- 
2.14.3.windows.1

