From ed7c244d039c4f8acdf05e3e44e96e216e225ff3 Mon Sep 17 00:00:00 2001
From: Jean Xiao <jianxiao@codeaurora.org>
Date: Tue, 19 May 2020 16:40:57 +0800
Subject: [PATCH] omxaacdec set audio DualMono config

---
 omx/gstomxaacdec.c   | 13 +++++++++++++
 omx/gstomxaudiodec.c |  1 +
 2 files changed, 14 insertions(+)

diff --git a/omx/gstomxaacdec.c b/omx/gstomxaacdec.c
index 66d881d..9100c05 100644
--- a/omx/gstomxaacdec.c
+++ b/omx/gstomxaacdec.c
@@ -162,6 +162,19 @@ gst_omx_aac_dec_set_format (GstOMXAudioDec * dec, GstOMXPort * port,
     return FALSE;
   }
 
+#ifdef USE_OMX_TARGET_QTI
+  QOMX_AUDIO_CONFIG_DUALMONOTYPE dualmono;
+  GST_OMX_INIT_STRUCT (&dualmono);
+  dualmono.eChannelConfig = OMX_AUDIO_DUAL_MONO_MODE_DEFAULT;
+  err =
+      gst_omx_component_set_config (dec->dec, QOMX_IndexConfigAudioDualMono,
+      &dualmono);
+  if (err != OMX_ErrorNone) {
+      GST_ERROR_OBJECT (self, "Error setting AAC config: QOMX_IndexConfigAudioDualMono (0x%08x)", err);
+      return FALSE;
+  }
+#endif
+
   return TRUE;
 }
 
diff --git a/omx/gstomxaudiodec.c b/omx/gstomxaudiodec.c
index d482391..eb3e977 100644
--- a/omx/gstomxaudiodec.c
+++ b/omx/gstomxaudiodec.c
@@ -350,6 +350,7 @@ gst_omx_audio_dec_loop (GstOMXAudioDec * self)
 
     GST_OMX_INIT_STRUCT (&pcm_param);
     pcm_param.nPortIndex = self->dec_out_port->index;
+    pcm_param.eEndian = OMX_EndianLittle;
     err =
         gst_omx_component_get_parameter (self->dec, OMX_IndexParamAudioPcm,
         &pcm_param);
-- 
2.17.1

