From 43338161250dabcb00bda19427072e7f7be5d64a Mon Sep 17 00:00:00 2001
From: Jean Xiao <jianxiao@codeaurora.org>
Date: Thu, 8 Aug 2019 16:21:38 +0800
Subject: [PATCH] meta-qti-gst: fix gstomx PortSettingsChanged event

---
 omx/gstomx.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/omx/gstomx.c b/omx/gstomx.c
index d7181bf..66ea698 100644
--- a/omx/gstomx.c
+++ b/omx/gstomx.c
@@ -569,6 +569,13 @@ EventHandler (OMX_HANDLETYPE hComponent, OMX_PTR pAppData, OMX_EVENTTYPE eEvent,
     }
     case OMX_EventPortSettingsChanged:
     {
+      //gstomx doesn't support OMX_IndexConfigCommonOutputCrop, ignore it
+      if (nData2 == OMX_IndexConfigCommonOutputCrop) {
+          GST_DEBUG_OBJECT (comp->parent, "%s ignore OMX_IndexConfigCommonOutputCrop (port index: %u)",
+                  comp->name, nData1);
+          break;
+      }
+
       GstOMXMessage *msg = g_slice_new (GstOMXMessage);
       OMX_U32 index;

--
1.9.1

