diff -uprN a/config/bellagio/gstomx.conf b/config/bellagio/gstomx.conf
--- a/config/bellagio/gstomx.conf	2018-09-13 19:40:53.037836000 +0800
+++ b/config/bellagio/gstomx.conf	2018-09-13 18:36:06.411511000 +0800
@@ -32,4 +32,22 @@ component-name=OMX.qcom.audio.decoder.am
 rank=0
 in-port-index=0
 out-port-index=1
-hacks=no-disable-outport
\ No newline at end of file
+hacks=no-disable-outport
+
+[omxh264dec]
+type-name=GstOMXH264Dec
+core-name=/usr/lib/libOmxCore.so
+component-name=OMX.qcom.video.decoder.avc
+rank=0
+in-port-index=0
+out-port-index=1
+hacks=no-disable-outport
+
+[omxh264enc]
+type-name=GstOMXH264Enc
+core-name=/usr/lib/libOmxCore.so
+component-name=OMX.qcom.video.encoder.avc
+rank=0
+in-port-index=0
+out-port-index=1
+hacks=no-disable-outport
