From 9cb1dd7b19342fa30606de46eb0ce18a56d0edde Mon Sep 17 00:00:00 2001
From: "Petar G. Georgiev" <petarg@codeaurora.org>
Date: Wed, 15 Apr 2020 10:11:07 +0100
Subject: [PATCH] omxh264enc: Add support for VUI timing info

---
 omx/gstomxh264enc.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/omx/gstomxh264enc.c b/omx/gstomxh264enc.c
index e46593a..838ddbf 100644
--- a/omx/gstomxh264enc.c
+++ b/omx/gstomxh264enc.c
@@ -624,6 +624,22 @@ gst_omx_h264_enc_set_format (GstOMXVideoEnc * enc, GstOMXPort * port,
   }
 #endif
 
+#ifdef USE_OMX_TARGET_QTI
+  OMX_QCOM_VIDEO_PARAM_VUI_TIMING_INFO vui_timing_info;
+
+  GST_OMX_INIT_STRUCT(&vui_timing_info);
+  vui_timing_info.bEnable = OMX_TRUE;
+  err =
+      gst_omx_component_set_parameter (GST_OMX_VIDEO_ENC (self)->enc,
+      OMX_QcomIndexParamH264VUITimingInfo, &vui_timing_info);
+  if (err != OMX_ErrorNone) {
+    GST_ERROR_OBJECT (self,
+        "can't set OMX_QcomIndexParamH264VUITimingInfo %s (0x%08x)",
+        gst_omx_error_to_string (err), err);
+    return FALSE;
+  }
+#endif
+
   /* Configure GOP pattern */
   if (self->periodicty_idr !=
       GST_OMX_H264_VIDEO_ENC_PERIODICITY_OF_IDR_FRAMES_DEFAULT
-- 
2.17.1

