diff -Naur a/omx/gstomx.h b/omx/gstomx.h
--- a/omx/gstomx.h 2018-10-15 15:46:32.106715906 +0800
+++ b/omx/gstomx.h 2018-10-15 15:58:43.383341181 +0800
@@ -372,6 +372,16 @@
 OMX_ERRORTYPE     gst_omx_port_wait_enabled (GstOMXPort * port, GstClockTime timeout);
 gboolean          gst_omx_port_is_enabled (GstOMXPort * port);

+/** Define the OMX IL version that corresponds to this set of header files.
+ *  We also define a combined version that can be used to write or compare
+ *  values of the 32bit nVersion field, assuming a little endian architecture */
+#define OMX_VERSION_MAJOR 1
+#define OMX_VERSION_MINOR 1
+#define OMX_VERSION_REVISION 2
+#define OMX_VERSION_STEP 0
+
+#define OMX_VERSION ((OMX_VERSION_STEP<<24) | (OMX_VERSION_REVISION<<16) | (OMX_VERSION_MINOR<<8) | OMX_VERSION_MAJOR)
+

 void              gst_omx_set_default_role (GstOMXClassData *class_data, const gchar *default_role);

