From ead30452dbce2c296b132d2c74340ce6f7854ac1 Mon Sep 17 00:00:00 2001
From: "Petar G. Georgiev" <quic_petarg@quicinc.com>
Date: Mon, 22 Aug 2022 19:17:27 +0300
Subject: [PATCH] omxvideoenc: Add support for ROI encoding on QTI target

Signed-off-by: Petar G. Georgiev <quic_petarg@quicinc.com>s
---
 omx/gstomxvideoenc.c | 263 +++++++++++++++++++++++++++++++++++++++++++
 omx/gstomxvideoenc.h |   3 +
 2 files changed, 266 insertions(+)

diff --git a/omx/gstomxvideoenc.c b/omx/gstomxvideoenc.c
index bce9774..011a768 100644
--- a/omx/gstomxvideoenc.c
+++ b/omx/gstomxvideoenc.c
@@ -94,6 +94,17 @@ gst_omx_video_enc_rotate_get_type (void)
   return qtype;
 }
 
+#ifdef USE_OMX_TARGET_QTI
+typedef struct _GstQuantRectangle GstQuantRectangle;
+struct _GstQuantRectangle {
+  gint x;
+  gint y;
+  gint w;
+  gint h;
+  gint qp;
+};
+#endif
+
 #ifdef USE_OMX_TARGET_ZYNQ_USCALE_PLUS
 #define GST_TYPE_OMX_VIDEO_ENC_QP_MODE (gst_omx_video_enc_qp_mode_get_type ())
 static GType
@@ -279,6 +290,9 @@ enum
   PROP_MAX_QUANT_P_FRAMES,
   PROP_MIN_QUANT_B_FRAMES,
   PROP_MAX_QUANT_B_FRAMES,
+  PROP_ROI_QUANT_MODE,
+  PROP_ROI_QUANT_META_VALUE,
+  PROP_ROI_QUANT_BOXES,
   PROP_NUM_LTR_FRAMES,
   PROP_NUM_HIER_LAYERS,
   PROP_QP_MODE,
@@ -322,6 +336,8 @@ enum
 #define GST_OMX_VIDEO_ENC_SLICE_SIZE_DEFAULT (0)
 #define GST_OMX_VIDEO_ENC_DEPENDENT_SLICE_DEFAULT (FALSE)
 #define GST_OMX_VIDEO_ENC_DEFAULT_ROI_QUALITY OMX_ALG_ROI_QUALITY_HIGH
+#define GST_OMX_VIDEO_ENC_ROI_QUANT_MODE_DEFAULT (FALSE)
+#define GST_OMX_VIDEO_ENC_ROI_QUANT_DELTA_DEFAULT (-15)
 #define GST_OMX_VIDEO_ENC_NUM_LTR_FRAMES_DEFAULT (0xffffffff)
 #define GST_OMX_VIDEO_ENC_NUM_HIER_LAYERS_DEFAULT (0xffffffff)
 
@@ -435,6 +451,38 @@ gst_omx_video_enc_class_init (GstOMXVideoEncClass * klass)
           G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
           GST_PARAM_MUTABLE_READY));
 
+  g_object_class_install_property (gobject_class, PROP_ROI_QUANT_MODE,
+      g_param_spec_boolean ("roi-quant-mode", "ROI Quantization Mode",
+          "Enable/Disable Adjustment of the quantization parameter according "
+          "to ROIs set manually via the 'roi-quant-boxes' property and/or "
+          "arriving as GstVideoRegionOfInterestMeta attached to the buffer",
+          GST_OMX_VIDEO_ENC_ROI_QUANT_MODE_DEFAULT,
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+
+  g_object_class_install_property (gobject_class, PROP_ROI_QUANT_META_VALUE,
+      g_param_spec_boxed ("roi-quant-meta-value", "ROI Meta Quantization Value",
+          "Set specific QP value, different then the default value of (-15), "
+          "for a GstVideoRegionOfInterestMeta type "
+          "(e.g. 'roi-meta-qp,person=-20,cup=10,dog=-5;'). The QP values must "
+          "be in the range of -31 (best quality) to 30 (worst quality)",
+          GST_TYPE_STRUCTURE,
+          G_PARAM_READWRITE| G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+
+  g_object_class_install_property (gobject_class, PROP_ROI_QUANT_BOXES,
+      gst_param_spec_array ("roi-quant-boxes", "ROI Quantization Boxes",
+          "Manually set ROI boxes (e.g. '<<X, Y, W, H, QP>, <X, Y, W, H, QP>>'). "
+          "The QP values must be in the range of -31 (best quality) to "
+          "30 (worst quality)",
+          gst_param_spec_array ("rectangle", "Rectangle", "Rectangle",
+              g_param_spec_int ("value", "Rectangle Value",
+                  "One of X, Y, WIDTH, HEIGHT or QP", G_MININT, G_MAXINT, 0,
+                  G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS),
+              G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS),
+          G_PARAM_READWRITE | G_PARAM_STATIC_STRINGS |
+          GST_PARAM_MUTABLE_READY));
+
   g_object_class_install_property (gobject_class, PROP_NUM_LTR_FRAMES,
       g_param_spec_uint ("num-ltr-frames", "LTR Frames Count",
           "Number of Long Term Reference Frames (0xffffffff=component default)",
@@ -622,6 +670,9 @@ gst_omx_video_enc_init (GstOMXVideoEnc * self)
   self->max_quant_p_frames = GST_OMX_VIDEO_ENC_MAX_QP_DEFAULT;
   self->min_quant_b_frames = GST_OMX_VIDEO_ENC_MIN_QP_DEFAULT;
   self->max_quant_b_frames = GST_OMX_VIDEO_ENC_MAX_QP_DEFAULT;
+  self->roi_quant_mode = GST_OMX_VIDEO_ENC_ROI_QUANT_MODE_DEFAULT;
+  self->roi_quant_values = gst_structure_new_empty ("roi-meta-qp");
+  self->roi_quant_boxes = g_array_new (FALSE, FALSE, sizeof (GstQuantRectangle));
   self->num_ltr_frames = GST_OMX_VIDEO_ENC_NUM_LTR_FRAMES_DEFAULT;
   self->num_hier_layers = GST_OMX_VIDEO_ENC_NUM_HIER_LAYERS_DEFAULT;
 #endif
@@ -1044,6 +1095,7 @@ gst_omx_video_enc_open (GstVideoEncoder * encoder)
     OMX_CONFIG_BOOLEANTYPE prepend_sps_pps_headers;
     OMX_QCOM_VIDEO_CONFIG_AUD audelimiter;
     OMX_QCOM_VIDEO_PARAM_IPB_QPRANGETYPE qprange;
+    OMX_BOOL enable;
 
     GST_OMX_INIT_STRUCT (&prepend_sps_pps_headers);
     prepend_sps_pps_headers.bEnabled = OMX_TRUE;
@@ -1109,6 +1161,25 @@ gst_omx_video_enc_open (GstVideoEncoder * encoder)
       return FALSE;
     }
 
+    enable = self->roi_quant_mode ? OMX_TRUE : OMX_FALSE;
+
+    err =
+        gst_omx_component_set_parameter (GST_OMX_VIDEO_ENC (self)->enc,
+        OMX_QTIIndexParamVideoEnableRoiInfo, &enable);
+
+    if (err == OMX_ErrorUnsupportedIndex) {
+      GST_WARNING_OBJECT (self,
+          "Setting OMX_QTIIndexParamVideoEnableRoiInfo not supported by the component");
+    } else if (err == OMX_ErrorUnsupportedSetting) {
+      GST_WARNING_OBJECT (self,
+          "Setting ROI info not supported by the component");
+    } else if (err != OMX_ErrorNone) {
+      GST_ERROR_OBJECT (self,
+          "can't set OMX_QTIIndexParamVideoEnableRoiInfo %s (0x%08x)",
+          gst_omx_error_to_string (err), err);
+      return FALSE;
+    }
+
     if (self->num_ltr_frames != 0xffffffff) {
       QOMX_VIDEO_PARAM_LTRCOUNT_TYPE ltrframes;
 
@@ -1321,6 +1392,11 @@ gst_omx_video_enc_finalize (GObject * object)
   g_clear_pointer (&self->alg_roi_quality_enum_class, g_type_class_unref);
 #endif
 
+#ifdef USE_OMX_TARGET_QTI
+  g_array_free (self->roi_quant_boxes, TRUE);
+  gst_structure_free (self->roi_quant_values);
+#endif
+
   G_OBJECT_CLASS (gst_omx_video_enc_parent_class)->finalize (object);
 }
 
@@ -1385,6 +1461,52 @@ gst_omx_video_enc_set_property (GObject * object, guint prop_id,
     case PROP_MAX_QUANT_B_FRAMES:
       self->max_quant_b_frames = g_value_get_uint (value);
       break;
+    case PROP_ROI_QUANT_MODE:
+      self->roi_quant_mode = g_value_get_boolean (value);
+      break;
+    case PROP_ROI_QUANT_META_VALUE:
+      if (self->roi_quant_values)
+        gst_structure_free (self->roi_quant_values);
+
+      self->roi_quant_values = g_value_dup_boxed (value);
+      break;
+    case PROP_ROI_QUANT_BOXES:
+    {
+      guint idx = 0;
+
+      /* Remove all old values */
+      g_array_set_size (self->roi_quant_boxes, 0);
+
+      for (idx = 0; idx < gst_value_array_get_size (value); idx++) {
+        const GValue *v = gst_value_array_get_value (value, idx);
+        GstQuantRectangle qbox = { 0, 0, 0, 0, 0 };
+
+        if (gst_value_array_get_size (v) != 5) {
+          GST_WARNING_OBJECT (self,
+              "Invalid entries for ROI box at index '%u', ignoring", idx);
+          continue;
+        }
+
+        qbox.x = g_value_get_int (gst_value_array_get_value (v, 0));
+        qbox.y = g_value_get_int (gst_value_array_get_value (v, 1));
+        qbox.w = g_value_get_int (gst_value_array_get_value (v, 2));
+        qbox.h = g_value_get_int (gst_value_array_get_value (v, 3));
+        qbox.qp = g_value_get_int (gst_value_array_get_value (v, 4));
+
+        if (qbox.w == 0 || qbox.h == 0) {
+          GST_WARNING_OBJECT (self,
+              "Invalid dimensions for ROI box at index %u, ignoring", idx);
+          continue;
+        } else if ((qbox.qp < -31) || (qbox.qp > 30)) {
+          GST_WARNING_OBJECT (self,
+              "Invalid QP value for ROI box at index %u, ignoring", idx);
+          continue;
+        }
+
+        g_array_append_val (self->roi_quant_boxes, qbox);
+      }
+      break;
+    }
     case PROP_NUM_LTR_FRAMES:
       self->num_ltr_frames = g_value_get_uint (value);
       break;
@@ -1494,6 +1616,49 @@ gst_omx_video_enc_get_property (GObject * object, guint prop_id, GValue * value,
     case PROP_MAX_QUANT_B_FRAMES:
       g_value_set_uint (value, self->max_quant_b_frames);
       break;
+    case PROP_ROI_QUANT_MODE:
+      g_value_set_boolean (value, self->roi_quant_mode);
+      break;
+    case PROP_ROI_QUANT_META_VALUE:
+      if (self->roi_quant_values)
+        g_value_set_boxed (value, self->roi_quant_values);
+      break;
+    case PROP_ROI_QUANT_BOXES:
+    {
+      GstQuantRectangle *qbox = NULL;
+      guint idx = 0;
+
+      for (idx = 0; idx < self->roi_quant_boxes->len; idx++) {
+        GValue element = G_VALUE_INIT, val = G_VALUE_INIT;
+
+        g_value_init (&element, GST_TYPE_ARRAY);
+        g_value_init (&val, G_TYPE_INT);
+
+        qbox = &(g_array_index (self->roi_quant_boxes, GstQuantRectangle, idx));
+
+        g_value_set_int (&val, qbox->x);
+        gst_value_array_append_value (&element, &val);
+
+        g_value_set_int (&val, qbox->y);
+        gst_value_array_append_value (&element, &val);
+
+        g_value_set_int (&val, qbox->w);
+        gst_value_array_append_value (&element, &val);
+
+        g_value_set_int (&val, qbox->h);
+        gst_value_array_append_value (&element, &val);
+
+        g_value_set_int (&val, qbox->qp);
+        gst_value_array_append_value (&element, &val);
+
+        /* Append the rectangle to the output GST array */
+        gst_value_array_append_value (value, &element);
+
+        g_value_unset (&val);
+        g_value_unset (&element);
+      }
+      break;
+    }
     case PROP_NUM_LTR_FRAMES:
       g_value_set_uint (value, self->num_ltr_frames);
       break;
@@ -3236,6 +3401,100 @@ handle_roi_metadata (GstOMXVideoEnc * self, GstBuffer * input)
 }
 #endif
 
+#ifdef USE_OMX_TARGET_QTI
+static void
+handle_roi_metadata (GstOMXVideoEnc * self, GstBuffer * input)
+{
+  GstMeta *meta = NULL;
+  gpointer state = NULL;
+  OMX_QTI_VIDEO_CONFIG_ROI_RECT_REGION_INFO roiparam;
+  gint idx = 0, qpdelta = 0;
+
+  /* ROI mode is disabled, nothing to do except to return immediately */
+  if (!self->roi_quant_mode)
+    return;
+
+  GST_OMX_INIT_STRUCT (&roiparam);
+
+  while ((meta =
+          gst_buffer_iterate_meta_filtered (input, &state,
+              GST_VIDEO_REGION_OF_INTEREST_META_API_TYPE))) {
+    GstVideoRegionOfInterestMeta *roi = (GstVideoRegionOfInterestMeta *) meta;
+    GstStructure *s = NULL;
+    const gchar *label = NULL;
+
+    if (MAX_RECT_ROI_NUM == roiparam.nRegionNum) {
+      GST_WARNING_OBJECT (self,
+          "Received more than the allowed ROI, clipping to %d!",
+          MAX_RECT_ROI_NUM);
+      break;
+    }
+
+    if (roi->roi_type != g_quark_from_static_string ("ObjectDetection"))
+      continue;
+
+    s = gst_video_region_of_interest_meta_get_param (roi, "ObjectDetection");
+    label = gst_structure_get_string (s, "label");
+
+    GST_LOG_OBJECT (self, "Input buffer ROI: label=%s id=%d (%d, %d) %dx%d",
+        label, roi->id, roi->x, roi->y, roi->w, roi->h);
+
+    roiparam.nRegions[roiparam.nRegionNum].nLeft = roi->x;
+    roiparam.nRegions[roiparam.nRegionNum].nTop = roi->y;
+    roiparam.nRegions[roiparam.nRegionNum].nRight = roi->x + roi->w;
+    roiparam.nRegions[roiparam.nRegionNum].nBottom = roi->y + roi->h;
+
+    if (gst_structure_has_field (self->roi_quant_values, label)){
+      if (gst_structure_get_int (self->roi_quant_values, label, &qpdelta) &&
+          (qpdelta > -31) && (qpdelta < 30)) {
+        GST_LOG_OBJECT (self, "Use encoding QP delta (%d) for '%s'",
+            qpdelta, label);
+      } else {
+        qpdelta = GST_OMX_VIDEO_ENC_ROI_QUANT_DELTA_DEFAULT;
+        GST_WARNING_OBJECT (self,"Invalid QP delta for '%s', use default (%d)",
+            label, qpdelta);
+      }
+    } else {
+      qpdelta = GST_OMX_VIDEO_ENC_ROI_QUANT_DELTA_DEFAULT;
+      GST_LOG_OBJECT (self, "No QP delta specified for '%s', use default (%d)",
+          label, qpdelta);
+    }
+
+    roiparam.nRegions[roiparam.nRegionNum].nDeltaQP = qpdelta;
+    roiparam.nRegionNum++;
+  }
+
+  for (idx = 0; idx < self->roi_quant_boxes->len; idx++) {
+    GstQuantRectangle *qbox =
+        &(g_array_index (self->roi_quant_boxes, GstQuantRectangle, idx));
+
+    if (MAX_RECT_ROI_NUM == roiparam.nRegionNum) {
+      GST_WARNING_OBJECT (self,
+          "Received more than the allowed ROI, clipping to %d!",
+          MAX_RECT_ROI_NUM);
+      break;
+    }
+
+    GST_LOG_OBJECT (self, "Manual ROI: idx=%u (%d, %d) %dx%d with QP %d",
+        idx, qbox->x, qbox->y, qbox->w, qbox->h, qbox->qp);
+
+    roiparam.nRegions[roiparam.nRegionNum].nLeft = qbox->x;
+    roiparam.nRegions[roiparam.nRegionNum].nTop = qbox->y;
+    roiparam.nRegions[roiparam.nRegionNum].nRight = qbox->x + qbox->w;
+    roiparam.nRegions[roiparam.nRegionNum].nBottom = qbox->y + qbox->h;
+    roiparam.nRegions[roiparam.nRegionNum].nDeltaQP = qbox->qp;
+    roiparam.nRegionNum++;
+  }
+
+  roiparam.nPortIndex = self->enc_in_port->index;
+  GST_OMX_SET_TICKS (roiparam.nTimeStamp,
+      gst_util_uint64_scale (input->pts, OMX_TICKS_PER_SECOND, GST_SECOND));
+
+  gst_omx_component_set_config (self->enc,
+      (OMX_INDEXTYPE) OMX_QTIIndexConfigVideoRoiRectRegionInfo, &roiparam);
+}
+#endif
+
 static GstFlowReturn
 gst_omx_video_enc_handle_frame (GstVideoEncoder * encoder,
     GstVideoCodecFrame * frame)
@@ -3429,6 +3688,10 @@ gst_omx_video_enc_handle_frame (GstVideoEncoder * encoder,
     handle_roi_metadata (self, frame->input_buffer);
 #endif
 
+#ifdef USE_OMX_TARGET_QTI
+    handle_roi_metadata (self, frame->input_buffer);
+#endif
+
     /* Copy the buffer content in chunks of size as requested
      * by the port */
     if (fill_buffer
diff --git a/omx/gstomxvideoenc.h b/omx/gstomxvideoenc.h
index b066285..6d1b616 100644
--- a/omx/gstomxvideoenc.h
+++ b/omx/gstomxvideoenc.h
@@ -83,6 +83,9 @@ struct _GstOMXVideoEnc
   guint32 max_quant_p_frames;
   guint32 min_quant_b_frames;
   guint32 max_quant_b_frames;
+  gboolean roi_quant_mode;
+  GstStructure *roi_quant_values;
+  GArray *roi_quant_boxes;
   guint32 num_ltr_frames;
   guint32 num_hier_layers;
 #endif
-- 
2.17.1

