From 6babe5d396bf7cc7f797d3709df52710b508ec76 Mon Sep 17 00:00:00 2001
From: skomakul <quic_skomakul@quicinc.com>
Date: Wed, 19 Oct 2022 15:06:17 +0530
Subject: [PATCH] waylandsink: Release buffer on GST_QUERY_DRAIN

---
 ext/wayland/gstwaylandsink.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/ext/wayland/gstwaylandsink.c b/ext/wayland/gstwaylandsink.c
index 8fb4899..127b0aa 100644
--- a/ext/wayland/gstwaylandsink.c
+++ b/ext/wayland/gstwaylandsink.c
@@ -106,6 +106,8 @@ static GstCaps *gst_wayland_sink_get_caps (GstBaseSink * bsink,
 static gboolean gst_wayland_sink_set_caps (GstBaseSink * bsink, GstCaps * caps);
 static gboolean
 gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query);
+static gboolean
+gst_wayland_sink_query (GstBaseSink * bsink, GstQuery * query);
 static gboolean gst_wayland_sink_show_frame (GstVideoSink * bsink,
     GstBuffer * buffer);
 
@@ -204,6 +206,7 @@ gst_wayland_sink_class_init (GstWaylandSinkClass * klass)
   gstbasesink_class->set_caps = GST_DEBUG_FUNCPTR (gst_wayland_sink_set_caps);
   gstbasesink_class->propose_allocation =
       GST_DEBUG_FUNCPTR (gst_wayland_sink_propose_allocation);
+  gstbasesink_class->query = GST_DEBUG_FUNCPTR (gst_wayland_sink_query);
 
   gstvideosink_class->show_frame =
       GST_DEBUG_FUNCPTR (gst_wayland_sink_show_frame);
@@ -715,6 +718,24 @@ gst_wayland_sink_propose_allocation (GstBaseSink * bsink, GstQuery * query)
   return TRUE;
 }
 
+static gboolean
+gst_wayland_sink_query (GstBaseSink * bsink, GstQuery * query)
+{
+  GstWaylandSink *sink = GST_WAYLAND_SINK (bsink);
+
+  if (GST_QUERY_TYPE (query) == GST_QUERY_DRAIN) {
+    /*last_buffer will be held by compositor and will be
+    released when next buffer is received.
+    to release the original buffer back to the upstream buffer
+    pool without affecting the compositor we do last_buffer copy.*/
+    GstBuffer *buffer = gst_buffer_copy_deep (sink->last_buffer);
+    gst_buffer_replace (&sink->last_buffer, buffer);
+    gst_wl_display_force_release_buffers (sink->display);
+  }
+
+  return GST_BASE_SINK_CLASS (parent_class)->query (bsink, query);
+}
+
 static void
 frame_redraw_callback (void *data, struct wl_callback *callback, uint32_t time)
 {
